src/core/reference/src/utils/phillox_converter.cpp

```diff
@@ -0,0 +1,513 @@
+    convert_types out_val = {((static_cast<uint64_t>(1023) << 52) | significant)};
+    return out_val.d - 1.0;
+}
+
+// Helper function for converting uint32 values to bfloat16. Sets fractional part of
+// floating value with bits from uint32 value. Resulting value is in interval [0,1).
+bfloat16 uint32_to_bfloat16(uint32_t x) {
+    // bfloat16 is formatted as follows: sign(1 bit) exponent(8 bits) mantissa(7 bits). The value is interpreted
+    // The value is interpreted using following formula:
+    // (-1)^sign * 1, mantissa * 2 ^ (exponent - 127)
+    // Here we set the following values:
+    // sign = 0
+    // exponent = 127, for obtaining a zero exponent.
+    // mantissa = 7 right bits from generated uint32 random value.
+
+    auto x_uint16 = static_cast<uint16_t>(x);
+    convert_types out_val = {(static_cast<uint16_t>(127) << 7) | (x_uint16 & 0x7fu)};
+    return out_val.bf16 - static_cast<bfloat16>(1);
+}
+
+// Helper function for converting uint32 values to any T type float, using PyTorch
+// conversion method. Sets fractional part of floating value with bits from
+// uint32 value. Resulting value is in interval [0,1).
+template <typename T, typename V>
+PytorchAccumulatorType<T> uint32_to_T_type_float(V x) {
+    const auto mask = static_cast<V>((uint64_t(1) << std::numeric_limits<T>::digits) - 1);
+    const auto divisor = static_cast<PytorchAccumulatorType<T>>(1) / (uint64_t(1) << std::numeric_limits<T>::digits);
+    PytorchAccumulatorType<T> ret = (x & mask) * divisor;
+    return ret;
+}
+
+float uint32_to_bfloat16_pytorch(uint32_t x) {
+    const auto mask = static_cast<uint32_t>((1UL << 8) - 1);
+    const auto divisor = static_cast<PytorchAccumulatorType<ov::bfloat16>>(1) / (1UL << 8);
+    PytorchAccumulatorType<ov::bfloat16> ret = (x & mask) * divisor;
+    return ret;
+}
+
+// Converts uint32 values to destination type and normalizes to required range.
+template <typename T>
+void convert_to_output_type(const std::vector<uint32_t>& res,
+                            size_t uints_generated_per_exec,
+                            size_t output_type_generated_values,
+                            const element::Type& elem_type,
+                            const char* min_val,
+                            const char* max_val,
+                            char* out,
+                            size_t k,
+                            size_t elem_count,
+                            T (*convert_single_input)(uint32_t) = nullptr,
+                            T (*convert_two_inputs)(uint32_t, uint32_t, T, T) = nullptr,
+                            T (*mod_func)(uint32_t, T, T) = nullptr) {
+    // Get min and max values
+    T mn[1];
+    T mx[1];
+    memcpy(mn, min_val, elem_type.size());
+    memcpy(mx, max_val, elem_type.size());
+
+    std::vector<T> res_out_type(uints_generated_per_exec);
+    // Either 2 or 4 (2 for 64 bit values, 4 otherwise)
+    if (output_type_generated_values == 2) {
```

Comment:
Could you please verify and explain initialization of `res_out_type` vector with size `uints_generated_per_exec` instead of `output_type_generated_values`?

Reply:
Honestly I have no idea. Replaced with the output_type_generated_values